pkg/minikube/cluster/cluster_provisioner.go

```diff
@@ -0,0 +1,293 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This provisioner only for RHEL/CentOS iso which created to consume
+// with minishift
+
+package cluster
+
+import (
+	"bytes"
+	"bufio"
+	"errors"
+	"fmt"
+	"text/template"
+	"strings"
+	"path"
+	"os"
+	"syscall"
+
+	"github.com/docker/machine/libmachine/auth"
+	"github.com/docker/machine/libmachine/drivers"
+	"github.com/docker/machine/libmachine/engine"
+	"github.com/docker/machine/libmachine/log"
+	"github.com/docker/machine/libmachine/mcnutils"
+	"github.com/docker/machine/libmachine/provision/pkgaction"
+	"github.com/docker/machine/libmachine/swarm"
+	"github.com/docker/machine/libmachine/provision"
+
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+var (
+	engineConfigTemplateRHEL = `[Unit]
+Description=Docker Application Container Engine
+After=network.target rc-local.service
+Requires=rhel-push-plugin.socket
+
+[Service]
+Type=notify
+ExecStart=/usr/bin/docker daemon -H tcp://0.0.0.0:{{.DockerPort}} -H unix:///var/run/docker.sock \
+           --authorization-plugin rhel-push-plugin
+           --storage-driver devicemapper --tlsverify --tlscacert {{.AuthOptions.CaCertRemotePath}} \
+           --tlscert {{.AuthOptions.ServerCertRemotePath}} --tlskey {{.AuthOptions.ServerKeyRemotePath}} \
+           {{ range .EngineOptions.Labels }}--label {{.}} {{ end }}{{ range .EngineOptions.InsecureRegistry }}--insecure-registry {{.}} {{ end }}{{ range .EngineOptions.RegistryMirror }}--registry-mirror {{.}} {{ end }}{{ range .EngineOptions.ArbitraryFlags }}--{{.}} {{ end }}
+ExecReload=/bin/kill -s HUP $MAINPID
+MountFlags=slave
+LimitNOFILE=infinity
+LimitNPROC=infinity
+LimitCORE=infinity
+TimeoutStartSec=0
+Delegate=yes
+KillMode=process
+Environment={{range .EngineOptions.Env}}{{ printf "%q" . }} {{end}}
+
+[Install]
+WantedBy=multi-user.target
+`
+	engineConfigTemplateCentOS = `[Unit]
+Description=Docker Application Container Engine
+After=network.target rc-local.service
+
+[Service]
+Type=notify
+ExecStart=/usr/bin/docker daemon -H tcp://0.0.0.0:{{.DockerPort}} -H unix:///var/run/docker.sock \
+           --storage-driver devicemapper --tlsverify --tlscacert {{.AuthOptions.CaCertRemotePath}} \
+           --tlscert {{.AuthOptions.ServerCertRemotePath}} --tlskey {{.AuthOptions.ServerKeyRemotePath}} \
+           {{ range .EngineOptions.Labels }}--label {{.}} {{ end }}{{ range .EngineOptions.InsecureRegistry }}--insecure-registry {{.}} {{ end }}{{ range .EngineOptions.RegistryMirror }}--registry-mirror {{.}} {{ end }}{{ range .EngineOptions.ArbitraryFlags }}--{{.}} {{ end }}
+ExecReload=/bin/kill -s HUP $MAINPID
+MountFlags=slave
+LimitNOFILE=infinity
+LimitNPROC=infinity
+LimitCORE=infinity
+TimeoutStartSec=0
+Delegate=yes
+KillMode=process
+Environment={{range .EngineOptions.Env}}{{ printf "%q" . }} {{end}}
+
+[Install]
+WantedBy=multi-user.target
+`
+)
+
+func init() {
+	provision.Register("minishift", &provision.RegisteredProvisioner{
```

Comment:
As mentioned before, I have a hard time seeing how provision.Register is going to work here (w/o a custom Detector). This will just add the provisioner to the existing list of providioners. Given that we already have a CentOS/RHEL provisioner, how is this ensuring that the right provisoner is run.

I think to make this deterministic we need a custom Detector which primarily at the os release it. If it is minishift, it selects either out custom CentOS resp. RHEL provisoner. If it is not minishift it delegates to the StandardDetector which will just deal with the provisoners it knows. 

I think _RegisteredProvisioner_ is for adding provisioners for new OS types. Once you want to replace an existing one, you need a Detector as well.